vault/seal_autoseal.go

```diff
@@ -499,3 +512,76 @@ func (d *autoSeal) migrateRecoveryConfig(ctx context.Context) error {
 
 	return nil
 }
+
+// StartHealthCheck starts a goroutine that tests the health of the auto-unseal backend once every 10 minutes.
+// If unhealthy, logs a warning on the condition and begins testing every one minute until healthy again.
+func (d *autoSeal) StartHealthCheck() {
+	d.StopHealthCheck()
+	d.hcLock.Lock()
+	defer d.hcLock.Unlock()
+
+	healthCheck := time.NewTicker(sealHealthTestIntervalNominal)
+	d.healthCheckStop = make(chan struct{})
+	healthCheckStop := d.healthCheckStop
+	go func() {
+		lastTestOk := true
+		lastSeenOk := time.Now()
+		for {
+			select {
+			case <-healthCheckStop:
+				if healthCheck != nil {
+					healthCheck.Stop()
+				}
+				healthCheckStop = nil
+				return
+			case t := <-healthCheck.C:
+				testVal := fmt.Sprintf("Heartbeat %d", mathrand.Intn(1000))
+				ciphertext, err := d.Access.Encrypt(d.core.activeContext, []byte(testVal), nil)
```

Comment:
I think there might still be a race here?  If we're sealing, our stop channel gets closed, but sealing could complete and Core.activeContext changed before we see the closed channel.  It's unlikely, but I think either StopHealthCheck should wait for the goroutine to exit, or we should define a local ctx var to hold activeContext before starting the for loop.

Reply:
Hmm, yeah, remote but possible.